Source/Core/Core/IOS/Crypto/Sha.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include "Core/IOS/Crypto/Sha.h"
+
+#include <algorithm>
+#include <iterator>
+#include <mbedtls/sha1.h>
+#include <memory>
+
+#include "Common/Assert.h"
+#include "Common/ChunkFile.h"
+#include "Common/CommonTypes.h"
+#include "Core/HW/MMIO.h"
+#include "Core/HW/Memmap.h"
+#include "Core/System.h"
+
+namespace IOS::HLE
+{
+ShaDevice::ShaDevice(EmulationKernel& ios, const std::string& device_name)
+    : EmulationDevice(ios, device_name)
+{
+}
+
+std::optional<IPCReply> ShaDevice::Open(const OpenRequest& request)
+{
+  return Device::Open(request);
+}
+
+void ConvertContext(ShaDevice::ShaContext& src, mbedtls_sha1_context& dest)
+{
+  std::copy(std::begin(src.Length), std::end(src.Length), std::begin(dest.total));
+  std::copy(std::begin(src.States), std::end(src.States), std::begin(dest.state));
+}
+
+void ConvertContext(mbedtls_sha1_context& src, ShaDevice::ShaContext& dest)
+{
+  std::copy(std::begin(src.total), std::end(src.total), std::begin(dest.Length));
+  std::copy(std::begin(src.state), std::end(src.state), std::begin(dest.States));
+}
```

Comment:
These functions shouldn't be visible from outside the compilation unit so they should be `static`. `src` should be `const`. Also consider making `dest` a pointer so the direction of the conversion is more obvious at the call site.

Reply:
Yes, that's what [our style guide](https://github.com/dolphin-emu/dolphin/blob/master/Contributing.md#functions) would want. (We don't recommend raw pointers created using `new`, but a raw pointer as a function parameter where there aren't any ownership considerations is fine).